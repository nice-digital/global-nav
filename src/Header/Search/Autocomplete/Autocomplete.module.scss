@import '~@nice-digital/nds-core/scss/core';

/* stylelint-disable */

@import '../../header-settings';

.ac {
  :global {
    .autocomplete__wrapper {
      position: relative;
    }

    .autocomplete__hint,
    .autocomplete__input {
      -webkit-appearance: none;
      border: 1px solid $nds-colour-nice-black;
      border-radius: 0; /* Safari 10 on iOS adds implicit border rounding. */
      box-sizing: border-box;
      margin-bottom: 0; /* BUG: Safari 10 on macOS seems to add an implicit margin. */
      width: 100%;
    }

    .autocomplete__input {
      // The importants here are to override conflicting global styles in niceorg
      background-color: $nds-colour-nice-grey-3 !important;
      height: rem($search-height) !important;
      margin: 0 !important;
      max-width: 100% !important;
      position: relative;
      z-index: $base-z-index + 1;

      &:focus {
        outline: 3px solid $nds-colour-focus;
        outline-offset: -3px;
        z-index: $base-z-index - 1;
      }

      &::placeholder {
        color: #666 !important;
        font-size: 100%;
        font-weight: normal;
        opacity: 1; // Chrome uses a semi-transparent placeholder by default
      }
    }

    .autocomplete__hint {
      color: #bfc1c3;
      position: absolute;
    }

    .autocomplete__input--default {
      padding: rem($nds-spacing-small);
    }

    .autocomplete__input--focused {
      outline: 3px solid $nds-colour-focus;
      outline-offset: -3px;
    }

    .autocomplete__input--show-all-values {
      padding: 4px 34px 4px 4px;
      cursor: pointer;
    }

    .autocomplete__dropdown-arrow-down {
      z-index: -1;
      display: inline-block;
      position: absolute;
      right: 8px;
      width: 24px;
      height: 24px;
      top: 10px;
    }

    .autocomplete__menu {
      background-color: #fff;
      border: 1px solid $nds-colour-nice-grey-2;
      border-top: 0;
      color: #34384b;
      margin: 0;
      max-width: 100%;
      padding: 0;
      text-align: left;
      width: calc(100vw - #{2 * rem($nds-container-gutter)});
      @include mq($from: sm) {
        width: 100%;
      }
    }

    .autocomplete__menu--visible {
      display: block;
    }

    .autocomplete__menu--hidden {
      display: none;
    }

    .autocomplete__menu--overlay {
      box-shadow: rem(
          0 $nds-spacing-x-small $nds-spacing-medium $nds-spacing-x-small
        )
        rgba(0, 0, 0, 0.1);
      left: 0;
      position: absolute;
      top: 100%;
      z-index: $base-z-index;
    }

    .autocomplete__menu--inline {
      position: relative;
    }

    .autocomplete__option {
      border-bottom: solid $nds-colour-nice-grey-2;
      border-width: 1px 0;
      color: $nds-colour-link;
      cursor: pointer;
      display: block;
      margin: 0;
      position: relative;

      a {
        color: inherit;
        display: block;
        font-weight: normal;
        padding: rem($nds-spacing-small);
        text-decoration: none;
      }

      mark {
        background: 0;
        color: inherit;
        font-weight: bold;
        padding: 0;
      }
    }

    .autocomplete__option > * {
      pointer-events: none;
    }

    .autocomplete__option:first-of-type {
      border-top-width: 0;
    }

    .autocomplete__option:last-of-type {
      border-bottom-width: 0;
    }

    .autocomplete__option--odd {
      background-color: #fafafa;
    }

    .autocomplete__option--focused,
    .autocomplete__option:hover {
      background-color: $nds-colour-nice-grey-3;
      color: $nds-colour-link;
      outline: none;
      text-decoration: underline;
    }

    .autocomplete__option--no-results {
      background-color: #fafafa;
      color: #646b6f;
      cursor: not-allowed;
    }

    .autocomplete__hint,
    .autocomplete__input,
    .autocomplete__option {
      font-size: 16px;
      line-height: 1.25;
    }

    .autocomplete__hint,
    .autocomplete__option {
      //padding: 4px;
    }
  }
}
