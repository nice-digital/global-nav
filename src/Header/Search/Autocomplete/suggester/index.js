import { tokenize } from "./tokenizer";
import { findPhrase, highlightPhrase } from "./phrase-finder";
import { findTerms, highlightTerm } from "./term-finder";

const isValidSuggestion = function (suggestion) {
	return suggestion.phraseMatch != null || suggestion.termMatches.length > 0;
};

const tokenizeSuggestion = function (suggestion) {
	// TODO: Cache tokenized suggestions for performance
	return Object.assign({}, suggestion, {
		tokens: tokenize(suggestion.Title),
	});
};

const highlightSuggestion = function (suggestion) {
	let titleHtml;

	// Only highlight phrases, or individual terms, not both
	if (suggestion.phraseMatch)
		titleHtml = highlightPhrase(suggestion.Title, suggestion.phraseMatch);
	else
		titleHtml = suggestion.termMatches
			.sort(function (a, b) {
				return (
					// Order by later matches first so we can replace highlight matched terms
					b.suggestionToken.originalIndex - a.suggestionToken.originalIndex
				);
			})
			.reduce(function (titleHtml, termMatch) {
				return highlightTerm(titleHtml, termMatch);
			}, suggestion.Title);

	return Object.assign({}, suggestion, {
		titleHtml: titleHtml,
	});
};

const sortSuggestions = function (a, b) {
	// Show phrases first, and phrases closer to the start of the string first
	if (a.phraseMatch && b.phraseMatch) {
		return (
			a.phraseMatch.suggestionTokens[0].originalIndex -
			b.phraseMatch.suggestionTokens[0].originalIndex
		);
	}

	// Prefer phrases over term matches
	if (a.phraseMatch && !b.phraseMatch) return -1;
	if (!a.phraseMatch && b.phraseMatch) return 1;

	// Show term matches towards the start of the string first
	return (
		a.termMatches[0].suggestionToken.originalIndex -
		b.termMatches[0].suggestionToken.originalIndex
	);
};

/**
 *
 * @param {Array<{Title, Link}>} suggestions The complete list of sugggestions to filter
 * @param {String} query The search query
 * @param {Number} maxResults The maximum number of results to show
 */
export const suggester = function (suggestions, query, maxResults = 5) {
	const queryTokens = tokenize(query ? query.trim() : "");

	const findPhrasesAndTerms = function (suggestion) {
		const phraseMatch = findPhrase(queryTokens, suggestion.tokens),
			// Only look for individual tokens if there are no phrases
			termMatches = phraseMatch
				? []
				: findTerms(queryTokens, suggestion.tokens);

		return Object.assign({}, suggestion, {
			phraseMatch: phraseMatch,
			termMatches: termMatches,
		});
	};

	return suggestions
		.map(tokenizeSuggestion)
		.map(findPhrasesAndTerms)
		.filter(isValidSuggestion)
		.map(highlightSuggestion)
		.sort(sortSuggestions)
		.slice(0, maxResults)
		.map(function ({ Title, Link, TypeAheadType, titleHtml }) {
			// keep only the properties we care about
			return {
				Title,
				Link,
				TitleHtml: titleHtml,
				TypeAheadType,
			};
		});
};
