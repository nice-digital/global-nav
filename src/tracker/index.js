export const eventTimeout = 2000;

export const eventName = "GlobalNav";
export const defaultEventCategory = "TopHat and footer";
export const headerClickEventAction = "Tophat click";
export const footerClickEventAction = "Footer click";

export const ensureDataLayer = function () {
	window.dataLayer = window.dataLayer || [];
};

/**
 * Pushes an event object to the GTM dataLayer
 *
 * @param {string} category The event category
 * @param {string} action The event action
 * @param {string} label The event label
 * @param {Number} value The optional event value
 * @param {string} destinationUrl The optional link destination
 * @param {Function} callback The callback function to be called when the GTM event has fired
 */
export const trackEvent = function (
	category,
	action,
	label,
	value = null,
	destinationUrl = null,
	callback = null
) {
	ensureDataLayer();

	const dataLayerEvent = {
		event: eventName,
		eventCategory: category,
		eventAction: action,
		eventLabel: label,
	};

	if (value) {
		dataLayerEvent.eventValue = value;
	}

	if (destinationUrl) {
		dataLayerEvent.destinationUrl = destinationUrl;
	}

	if (callback) {
		let callbackTimeout;
		const wrappedCallback = function () {
			clearTimeout(callbackTimeout);
			callback();
		};
		callbackTimeout = setTimeout(wrappedCallback, eventTimeout);
		dataLayerEvent.eventCallback = wrappedCallback;
		dataLayerEvent.eventTimeout = eventTimeout;
	}

	window.dataLayer.push(dataLayerEvent);
};
