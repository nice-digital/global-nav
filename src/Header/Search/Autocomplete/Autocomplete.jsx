import React, { Component } from "react";
import AccessibleAutocomplete from "accessible-autocomplete/react";
import PropTypes from "prop-types";
import styles from "./Autocomplete.module.scss";

import { suggester } from "./suggester";
import { isIosDevice } from "./../../../utils";
import { trackEvent } from "./../../../tracker";

/**
 * Debounce
 * See http://unscriptable.com/2009/03/20/debouncing-javascript-methods/
 *
 * @param      {Function}  func       The function to debounce
 * @param      {Integer}  execAsap   Whether to execute the function now
 * @param      {Integer}  threshold  The detection period, in milliseconds
 * @param      {Object}  scope  The context for the debounced function
 * @return     {Function}  { The debounced function }
 */
const debounce = function (
	func,
	execAsap = false,
	threshold = 100,
	scope = null
) {
	let timeout;

	return function debounced() {
		let context = scope || this,
			args = arguments;

		function delayed() {
			if (!execAsap) func.apply(context, args);
			timeout = null;
		}

		if (timeout) clearTimeout(timeout);
		else if (execAsap) func.apply(context, args);

		timeout = setTimeout(delayed, threshold);
	};
};

/** The maximum number of autocomplete results to return */
const maxResults = 5;

/** Delay in millieconds before loading results */
export const rateLimitWait = 300;

const inputValueTemplate = function (suggestion) {
		if (!suggestion || !suggestion.Title) return "";
		return suggestion && suggestion.Title;
	},
	suggestionTemplateDefault = function (suggestion) {
		if (!suggestion || !suggestion.Link) return "";
		return `<a href="${suggestion.Link}">${
			suggestion.TitleHtml || suggestion.Title
		}</a>`;
	};

export default class Autocomplete extends Component {
	constructor(props) {
		super(props);

		this.state = {
			canUseDOM: false,
		};
	}

	componentDidMount() {
		this.setState({
			canUseDOM: true,
		});
	}

	onConfirm(suggestion) {
		if (suggestion) {
			var selectedEl = document.querySelectorAll(
				".autocomplete__option a[href='" + suggestion.Link + "']"
			);

			//This var is declared outside the eventcallback function to avoid
			//loosing the component scope due to hoisting
			// eslint-disable-next-line react/prop-types
			const { onNavigating } = this.props;
			const eventCallback = function () {
				const onNavigatingCallback =
					onNavigating &&
					(typeof onNavigating === "function"
						? onNavigating
						: window[onNavigating]);

				if (typeof onNavigatingCallback === "function") {
					onNavigatingCallback({
						element: selectedEl,
						href: suggestion.Link,
					});
				} else window.location.href = suggestion.Link;
			};

			if (suggestion.TypeAheadType) {
				trackEvent(
					"Search - Typeahead select",
					"Selected: " + suggestion.TypeAheadType,
					(
						suggestion.Title +
						" | " +
						document.getElementById("autocomplete").value
					).toLowerCase(),
					null,
					suggestion.Link,
					eventCallback
				);
			} else {
				trackEvent(
					"Search",
					"Typeahead select",
					suggestion.Title +
						" | " +
						document.getElementById("autocomplete").value,
					null,
					suggestion.Link,
					eventCallback
				);
			}
		}
	}

	suggest(query, populateResults) {
		if (
			this.props.source === false ||
			query === "" ||
			query === this.props.placeholder
		)
			return;

		let source;

		if (Array.isArray(this.props.source)) {
			source = this.props.source;
		}

		if (
			typeof this.props.source === "string" &&
			this.props.source.indexOf("/") === -1
		) {
			source = window[this.props.source];
			if (!source) return;
		}

		if (source) {
			populateResults(suggester(source, query, maxResults));
			return;
		}

		// Default to a URL for asynchronously loading suggestions from the server
		fetch(
			`${this.props.source}${
				this.props.source.indexOf("?") === -1 ? "?" : "&"
			}q=${query}`
		)
			.then((response) => response.json())
			.then((data) => {
				populateResults(data.slice(0, maxResults));
			});
	}

	render() {
		const templates = {
			inputValue: inputValueTemplate,
			suggestion: this.props.suggestionTemplate || suggestionTemplateDefault,
		};

		return (
			<div className={styles.ac}>
				{!this.props.source || !this.state.canUseDOM ? (
					<div className="autocomplete__wrapper">
						<input
							type="search"
							id="autocomplete"
							name="q"
							className="autocomplete__input autocomplete__input--default"
							placeholder={this.props.placeholder}
							defaultValue={this.props.query}
							data-hj-allow=""
							maxLength={512}
						/>
					</div>
				) : (
					<AccessibleAutocomplete
						id="autocomplete"
						name="q"
						placeholder={this.props.placeholder}
						displayMenu={isIosDevice() ? "inline" : "overlay"}
						minLength={3}
						source={debounce(this.suggest, false, rateLimitWait, this)}
						templates={templates}
						onConfirm={function (s) {
							this.onConfirm(s);
						}.bind(this)}
						confirmOnBlur={false}
						showNoOptionsFound={false}
						defaultValue={this.props.query}
						ref={function (acElement) {
							// TODO: This relies on an inner implementation detail of the autocomplete component, can we do this in a better way?
							var inputEl =
								acElement &&
								acElement.elementReferences &&
								acElement.elementReferences[-1];
							if (inputEl) {
								inputEl.setAttribute("data-hj-allow", "");
								inputEl.setAttribute("maxlength", 512);
							}
						}}
					/>
				)}
			</div>
		);
	}
}

Autocomplete.propTypes = {
	source: PropTypes.oneOfType([
		PropTypes.bool,
		PropTypes.string,
		PropTypes.arrayOf(
			PropTypes.shape({
				Title: PropTypes.string.isRequired,
				TitleHtml: PropTypes.string,
				TypeAheadType: PropTypes.string,
				Link: PropTypes.string.isRequired,
			})
		),
	]),
	suggestionTemplate: PropTypes.func,
	placeholder: PropTypes.string,
	query: PropTypes.string,
	onNavigating: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
};

Autocomplete.defaultProps = {
	source: false,
};
