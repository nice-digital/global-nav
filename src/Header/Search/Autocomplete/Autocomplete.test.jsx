import React from "react";
import { render, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Autocomplete, { rateLimitWait } from "./Autocomplete";

import { eventName, eventTimeout } from "./../../../tracker";

import { suggester } from "./suggester";

jest.mock("./suggester", () => ({
	suggester: jest.fn(jest.requireActual("./suggester").suggester),
}));

//let windowAssignMock = jest.fn();

//delete window.location;
//window.location = { assign: windowAssignMock };

describe("Autocomplete", () => {
	const defaultProps = {};

	window.location = { ...window.location, assign: jest.fn() };

	beforeEach(() => {
		window.dataLayer = [];

		suggester.mockClear();
	});

	afterEach(() => {
		window.location.assign.mockClear();
	});

	describe("Autocomplete", () => {
		it("should render autocomplete component", () => {
			const { container } = render(
				<Autocomplete {...defaultProps} source="/url" query="test" />
			);
			expect(container).toMatchSnapshot();
		});

		it("should render query into defaultValue attribute on autocomplete component", () => {
			const { getByRole } = render(
				<Autocomplete {...defaultProps} source="/url" query="diabetes" />
			);
			expect(getByRole("combobox")).toHaveValue("diabetes");
		});

		it("should add HotJar allowlist attribute to input box", () => {
			const { getByRole } = render(
				<Autocomplete {...defaultProps} source="/url" query="diabetes" />
			);

			expect(getByRole("combobox")).toHaveAttribute("data-hj-allow", "");
		});

		it("should add 512 character max length", () => {
			const { getByRole } = render(
				<Autocomplete {...defaultProps} source="/url" query="diabetes" />
			);

			expect(getByRole("combobox")).toHaveAttribute("maxlength", "512");
		});

		it("should use provided suggestion template", async () => {
			const option = {
				Title: "diabetes type 1",
				Link: "https://www.nice.org.uk/diabetes1.html",
			};

			const suggestionTemplate = jest.fn();

			const { getByRole } = render(
					<Autocomplete
						{...defaultProps}
						source={[option]}
						suggestionTemplate={suggestionTemplate}
					/>
				),
				input = getByRole("combobox"),
				user = userEvent.setup();

			await user.type(input, "dia");

			await waitFor(() => {
				expect(suggestionTemplate).toHaveBeenCalledTimes(1);
			});

			expect(suggestionTemplate.mock.calls[0][0]).toStrictEqual({
				...option,
				TitleHtml: "<mark>dia</mark>betes type 1",
				TypeAheadType: undefined,
			});
		});

		it("should push autocomplete select event to the dataLayer", async () => {
			const option = {
				Title: "diabetes type 1",
				Link: "https://www.nice.org.uk/diabetes1.html",
			};

			const { getByRole } = render(
					<Autocomplete {...defaultProps} source={[option]} query="dia" />
				),
				input = getByRole("combobox"),
				user = userEvent.setup();

			await user.type(input, "diab");

			await waitFor(async () => {
				const optionLink = getByRole("link", { name: "diab etes type 1" });
				await user.click(optionLink);
			});

			expect(window.dataLayer).toEqual([
				{
					event: eventName,
					eventCategory: "Search",
					eventAction: "Typeahead select",
					eventLabel: option.Title + " | diab",
					destinationUrl: option.Link,
					eventCallback: expect.any(Function),
					eventTimeout: eventTimeout,
				},
			]);
		});

		it("should push autocomplete select event to the dataLayer using TypeAheadType property", async () => {
			const option = {
				Title: "diabetes type 1",
				TypeAheadType: "keyword",
				Link: "https://www.nice.org.uk/diabetes1.html",
			};

			const { getByRole } = render(
					<Autocomplete {...defaultProps} source={[option]} query="dia" />
				),
				input = getByRole("combobox"),
				user = userEvent.setup();

			await user.type(input, "diab");

			await waitFor(async () => {
				const optionLink = getByRole("link", { name: "diab etes type 1" });
				await user.click(optionLink);
			});

			expect(window.dataLayer).toEqual([
				{
					event: eventName,
					eventCategory: "Search - Typeahead select",
					eventAction: "Selected: keyword",
					eventLabel: option.Title + " | diab",
					destinationUrl: option.Link,
					eventCallback: expect.any(Function),
					eventTimeout: eventTimeout,
				},
			]);
		});

		it("should navigate to selected option in event callback", async () => {
			const option = {
				Title: "diabetes type 1",
				Link: "https://www.nice.org.uk/diabetes1.html",
			};

			const { getByRole } = render(
					<Autocomplete {...defaultProps} source={[option]} query="dia" />
				),
				input = getByRole("combobox"),
				user = userEvent.setup();

			await user.type(input, "diab");

			await waitFor(async () => {
				const optionLink = getByRole("link", { name: "diab etes type 1" });
				await user.click(optionLink);
			});

			window.dataLayer[0].eventCallback();
			expect(window.location.assign).toHaveBeenCalledWith(
				"https://www.nice.org.uk/diabetes1.html"
			);
		});

		it("should call onNavigating prop function with selected option in event callback", async () => {
			const option = {
					Title: "diabetes type 1",
					Link: "/diabetes1.html",
				},
				onNavigating = jest.fn();

			const { getByRole } = render(
					<Autocomplete
						{...defaultProps}
						source={[option]}
						query="dia"
						onNavigating={onNavigating}
					/>
				),
				input = getByRole("combobox"),
				user = userEvent.setup();

			await user.type(input, "diab");

			await waitFor(async () => {
				const optionLink = getByRole("link", { name: "diab etes type 1" });
				await user.click(optionLink);
			});

			window.dataLayer[0].eventCallback();
			expect(onNavigating).toHaveBeenCalledWith({
				href: "/diabetes1.html",
				element: expect.anything(),
			});
		});

		it("should debounce suggestions", async () => {
			const option = {
				Title: "diabetes type 1",
				Link: "/diabetes1.html",
			};

			const { getByRole, queryAllByRole } = render(
					<Autocomplete {...defaultProps} source={[option]} />
				),
				input = getByRole("combobox"),
				user = userEvent.setup();

			await user.type(input, "dia");
			await user.type(input, "b");
			await user.type(input, "e");

			await waitFor(() => {
				const optionLinks = queryAllByRole("link");
				expect(optionLinks).toHaveLength(1);
			});

			expect(suggester).toBeCalledTimes(1);
			expect(suggester).toBeCalledWith([option], "diabe", 5);
		});
	});

	describe("Non-autocomplete", () => {
		it("should render fallback search box when no source", () => {
			const { container } = render(
				<Autocomplete {...defaultProps} source={false} query="diabetes" />
			);
			expect(container).toMatchSnapshot();
		});
	});
});
