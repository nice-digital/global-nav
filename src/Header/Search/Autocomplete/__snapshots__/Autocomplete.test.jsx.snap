// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete Autocomplete should render autocomplete component 1`] = `
<div>
  <div
    class="ac"
  >
    <label
      class="visually-hidden"
      data-headlessui-state=""
      id="headlessui-combobox-label-:r0:"
    />
    <input
      aria-autocomplete="list"
      aria-expanded="false"
      aria-labelledby="headlessui-combobox-label-:r0:"
      autocomplete="off"
      class="autocomplete__input"
      data-headlessui-state=""
      data-hj-allow=""
      id="autocomplete"
      maxlength="512"
      name="q"
      role="combobox"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`Autocomplete Non-autocomplete should render fallback search box when no source 1`] = `
<div>
  <div
    class="ac"
  >
    <div
      class="autocomplete__wrapper"
    >
      <input
        class="autocomplete__input autocomplete__input--default"
        data-hj-allow=""
        id="autocomplete"
        maxlength="512"
        name="q"
        type="search"
        value="diabetes"
      />
    </div>
  </div>
</div>
`;
